crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -815,29 +809,35 @@ where
     Ok(async move {
         let archiver = match maybe_archiver {
             Some(archiver) => archiver,
-            None => initialize_archiver(&segment_headers_store, &subspace_link, client.as_ref())?,
+            None => {
+                let Some(last_segment_header) = segment_headers_store.last_segment_header() else {
+                    unreachable!("Already checked above that max_segment_index is not None")
+                };
+                let last_archived_block = last_segment_header.last_archived_block();
+
+                // wait for unknown blocks to arrive.
+                while let Some(ref block_import_notification) =
+                    block_importing_notification_stream.next().await
+                {
+                    if block_import_notification
+                        .block_number
+                        .checked_sub(&(last_archived_block.number + 1).into())
+                        .is_none()
```

Comment:
First of all, checked subtraction is unnecessary complex, just use comparison between two numbers instead.

Second, we archive blocks at 100 blocks depth (`confirmation_depth_k`), so this is not actually correct and will skip more blocks than necessary, causing issues in the process. See how `block_number_to_archive` is derived below. Since we get notification right *before* the block is archived, we should stop this loop exactly one block before the block at which we'll resume archiving with the loop below.

Reply:
Well, actually no, it will work fine as is too, just not skipping as many blocks as it could otherwise. Let's at least replace the check with a simple comparison.

Reply:
>First of all, checked subtraction is unnecessary complex, just use comparison between two numbers instead.

Let me simply fix him, here it was actually just something wrong with my ide at the time he couldn't infer the type of block_import_notification.block_number, so I just copied the code below

>Second, we archive blocks at 100 blocks depth (confirmation_depth_k), so this is not actually correct and will skip more blocks than necessary, causing issues in the process. See how block_number_to_archive is derived below. Since we get notification right before the block is archived, we should stop this loop exactly one block before the block at which we'll resume archiving with the loop below.

I guess I should get confirmation_depth_k manually when I reach the edge, and then continue skipping?

Reply:
Currently it's actually split into two skips, first it reaches the edge and init_archiver gets confirmation_depth_k, then it continues to skip confirmation_depth_k and init_archiver again.  Really unnecessary initialization, I should have gotten him manually